geomstats/stiefel.py

```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        if n_tangent_vecs == 1:
+            tangent_vec = gs.tile(tangent_vec, (n_base_points, 1, 1))
+        n_retractions = gs.maximum(n_base_points, n_tangent_vecs)
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
+        # TODO: remove cycle, when qr will be vectorized
+        matrix_q = gs.zeros_like(base_point)
+        matrix_r = gs.zeros((n_retractions, p, p))
+
+        for i, k in enumerate(base_point + tangent_vec):
+            matrix_q[i], matrix_r[i] = gs.linalg.qr(k)
+
+        # flipping signs
+        # Q = gs.matmul(Q, gs.diag(gs.sign(gs.sign(gs.diagonal(R)) + 0.5)))
+        # TODO: remove cycle, when diag, diangonal will be vectorized
+        result = gs.zeros_like(base_point)
+        for i, _ in enumerate(matrix_r):
+            result[i] = gs.matmul(
+                matrix_q[i],
+                gs.diag(gs.sign(gs.sign(gs.diagonal(matrix_r[i])) + 0.5)))
+
+        return result
+
+    def lifting(self, point, base_point):
+        """
+        Lifting map, based on QR-decomposion:
+        P_x^{-1}(Q) = QR - X
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        n_points, _, _ = point.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, p, n = base_point.shape
+
+        assert (n_points == n_base_points
+                or n_points == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_points, 1, 1))
+        if n_points == 1:
+            point = gs.tile(point, (n_base_points, 1, 1))
+        n_liftings = gs.maximum(n_base_points, n_points)
+
+        def make_minor(i, matrix):
+            return matrix[:i+1, :i+1]
+
+        def make_column_r(i, matrix):
+            if i == 0:
+                if (matrix[0, 0] > 0):
+                    return gs.array([1. / matrix[0, 0]])
+                else:
+                    raise Exception("M[0,0] <= 0")
+            else:
+                return matrix[:i+1, i]
+
+        def make_b(i, matrix, list_matrices_r):
+            b = gs.ones(i+1)
+
+            for j in range(i):
+                b[j] = - gs.matmul(
+                    matrix[i, :j+1], list_matrices_r[j])
+
+            return b
+
+        matrix_r = gs.zeros((n_liftings, n, n))
+        matrix_m = gs.matmul(gs.transpose(base_point, axes=(0, 2, 1)), point)
+
+        for k in range(n_liftings):
+            columns_list = []
+            matrix_m_k = matrix_m[k]
+
+            # construct r_0
+            columns_list.append(make_column_r(0, matrix_m_k))
+
+            for i in range(1, n):
```

Comment:
can we explain why we start at 1 here? Isn't there a way to start from 0 and make things a bit cleaner?

Reply:
Done.